lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1995,6 +1995,13 @@ SelfAccessKindRequest::evaluate(Evaluator &evaluator, FuncDecl *FD) const {
     case AccessorKind::MutableAddress:
     case AccessorKind::Set:
     case AccessorKind::Modify:
+      // If we have a class-constrained protocol extension, where the
+      // protocol does not have a class requirement, then the setter is
+      // implicitly nonmutating, unless it has been explicitly marked as
+      // mutating.
+      if (AD->isEligibleToBeNonMutatingInsideClassConstrainedProtoExt())
+        return SelfAccessKind::NonMutating;
+
       if (AD->isInstanceMember() &&
           doesContextHaveValueSemantics(AD->getDeclContext()))
```

Comment:
Can you investigate why doesContextHaveValueSemantics() return true for a class-constrained extension? Can we fix the problem there instead?

Reply:
@slavapestov Basically, `doesContextHaveValueSemantics()` gets the DeclContext's declared interface type, which in this case is the extended protocol's type, which doesn't impose a class constraint so the method returns true.

Reply:
I think what we could do is to add another check to `doesContextHaveValueSemantics` where if the decl context is an extension then we check whether the extension's generic signature requires Self to be a class (i.e. the same check I added in this patch, but moved into `doesContextHaveValueSemantics()` instead).